### PR TITLE
[mds-policy-author][mds-geography-author] Extract geography author service from policy author.

### PR DESCRIPTION
**Pull request by janedotx**
## PR Checklist

 - [ ] simple searchable title - `[mds-db] Add PG env var`, `[config] Fix eslint config`
 - [ ] briefly describe the changes in this PR
 - [ ] mark as draft if should not be merged
 - [ ] write tests for all new functionality

## Impacts
- [ ] Provider
- [ ] Agency
- [ ] Audit
- [x] Policy
- [ ] Compliance
- [ ] Daily
- [ ] Native
- [ ] Policy Author

### CODE REVIEW

**Comment by janedotx** [comment]
Sequel to https://github.com/lacuna-tech/mds-core/pull/186 but with a clean history.

File: .vscode/settings.json [hunk 1]
```diff
@@ -6,12 +6,19 @@
...
-  "prettier.disableLanguages": ["markdown"],
+  "eslint.options": {
+    "ignorePath": ".gitignore"
+  },
+  "eslint.validate": [
+    "typescript"
+  ],
+  "prettier.disableLanguages": [
+    "markdown"
+  ],
```
Comment (drtyh2o) [inline .vscode/settings.json]: These all seem like unnecessary formatting changes.

File: .vscode/settings.json [hunk 2]
```diff
@@ -6,12 +6,19 @@
   "editor.formatOnSave": true,
   "editor.insertSpaces": true,
   "editor.tabSize": 2,
+  "workbench.editor.enablePreview": false,
```
Comment (drtyh2o) [inline .vscode/settings.json]: This should probably be in your `user` settings; not `workspace` settings.

File: bin/mdsctl [hunk 3]
```diff
@@ -137,7 +137,7 @@ declare -A apps=(
   ["mds-audit"]="https://localhost:[port]/audit,mds,app"
   ["mds-compliance"]="https://localhost:[port]/compliance,mds,app"
   ["mds-daily"]="https://localhost:[port]/daily,mds,app"
-  ["mds-jurisdiction"]="https://localhost:[port]/jurisdiction,mds,app"
+  ["mds-geography-author"]="https://localhost:[port]/geography-author,mds,app"
```
Comment (drtyh2o) [inline bin/mdsctl]: Looks like you have deleted some of the `jurisdiction` related changes.

File: container-images/mds-geography-author/index.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import { api } from '@mds-core/mds-geography-author'
+import { env } from '@container-images/env-inject'
+
+const { npm_package_name, npm_package_version, npm_package_git_commit, PORT = 4010 } = env()
+
+ApiServer(api).listen(PORT, () =>
+  /* eslint-reason avoids import of logger */
+  /* eslint-disable-next-line no-console */
+  console.log(`${npm_package_name} v${npm_package_version} (${npm_package_git_commit}) running on port ${PORT}`)
+)
```
Comment (drtyh2o) [inline container-images/mds-geography-author/index.ts]: The way these `ApiServer` modules get defined has changed ... check out recent examples; they now get wrapped in an `HttpServer`

File: packages/mds-geography-author/LICENSE [hunk 5]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (drtyh2o) [inline packages/mds-geography-author/LICENSE]: Looks like these `LICENSE` files are only in a few packages. Not sure why they are not either always necessary or not included in all. @karcass

File: packages/mds-geography-author/package.json [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    "@mds-core/mds-api-helpers": "0.1.26",
+    "@mds-core/mds-api-server": "0.1.26",
+    "@mds-core/mds-db": "0.1.26",
+    "@mds-core/mds-logger": "0.1.24",
+    "@mds-core/mds-providers": "0.1.26",
+    "@mds-core/mds-schema-validators": "0.1.2",
+    "@mds-core/mds-utils": "0.1.26",
+    "express": "4.17.1",
+    "sinon": "7.5.0",
+    "uuid": "3.3.3"
```
Comment (drtyh2o) [inline packages/mds-geography-author/package.json]: This is different than the version used everywhere else.

File: helm/mds/values.yaml.tpl [hunk 7]
```diff
@@ -59,6 +59,11 @@ apis:
     pathPrefix: /event-processor
     version: ${EVENT_PROCESSOR_VERSION}
     migration: false
+  mds-geography-author:
+    enabled: true
+    pathPrefix: /geography
```
[the remark this answers is not in the document]
Comment (janedotx) [reply helm/mds/values.yaml.tpl]: ah shit

Yes we are, I just want to get this PR out because not having `geography` isn't a big blocker.

File: packages/mds-geography-author/api.ts [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+      if (get_published) {
+        params.get_published = get_published === 'true'
+      }
+
+      if (get_unpublished) {
+        params.get_unpublished = get_unpublished === 'true'
+      }
+
+      /* If the user can only read published geos, and all they want is the unpublished geos,
+       * save the trip to the DB and return an array fast.
```
[the remark this answers is not in the document]
Comment (janedotx) [reply packages/mds-geography-author/api.ts]: sg

[the remark this answers is not in the document]
Comment (avatarneil) [reply packages/mds-geography-author/api.ts]: My argument is that we _should_ throw an error in both cases. Would make diagnosing why nothing is coming back for them much simpler (on both the client end and our end).

[the remark this answers is not in the document]
Comment (janedotx) [reply packages/mds-geography-author/api.ts]: Had a discussion with Neil, the behavior will become:
- silently filter if no params are supplied and scope is the more restricted one
- throw if params are supplied and scope is restricted

File: packages/mds-db/types.ts [hunk 9]
```diff
@@ -96,10 +96,12 @@ export interface VehicleEventCountResult {
 }
 
 export interface ReadGeographiesParams {
-  get_read_only: boolean
+  get_published?: boolean
+  get_unpublished?: boolean
+  geography_ids?: Array<UUID>
```
Comment (avatarneil) [inline packages/mds-db/types.ts]: Probably want this to be `UUID[]`, we don't use `Array<xxx>` in many other places.

File: packages/mds-geography-author/api.ts [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+    checkAccess(scopes => {
+      return scopes.includes('geographies:read:published') || scopes.includes('geographies:read:unpublished')
+    }),
+    async (req, res) => {
+      const { geography_id } = req.params
+      try {
+        const geography = await db.readSingleGeography(geography_id)
+        if (!geography.publish_date && !res.locals.scopes.includes('geographies:read:unpublished')) {
+          throw new InsufficientPermissionsError('permission to read unpublished geographies missing')
+        } else {
```
Comment (avatarneil) [inline packages/mds-geography-author/api.ts]: Surprised prettier doesn't try to remove the `else` case here, it's unnecessary. You can just have it follow the `if` without the extra syntax.

File: packages/mds-geography-author/package.json [hunk 11]
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "@mds-core/mds-geography-author",
+  "description": "MDS Geography Author API",
+  "version": "0.0.1",
+  "private": true,
+  "licence": "Apache-2.0",
+  "engineStrict": true,
+  "engines": {
+    "node": ">= 8.0.0",
```
Comment (avatarneil) [inline packages/mds-geography-author/package.json]: engines in here seems unnecessary (we don't have it in anything but policy/geo stuff FWIW) and we also definitely wouldn't want it to be a non-absolute value.

File: packages/mds-db/geographies.ts [hunk 12]
```diff
@@ -10,35 +10,53 @@ import { getReadOnlyClient, getWriteableClient } from './client'
...
+    if (get_published) {
       conditions.push(`publish_date IS NOT NULL`)
     }
 
+    if (get_unpublished) {
+      conditions.push(`publish_date IS NULL`)
+    }
+
+    if (geography_ids) {
+      const SQLified_geography_ids = geography_ids.map(id => {
```
Comment (macsj200) [inline packages/mds-db/geographies.ts]: @evanxadkins Does this trigger you?

Comment (macsj200) [reply packages/mds-db/geographies.ts]: We have this in another part of the codebase
```
(${Object.values(
        event
      )
        .map(val => `'${val}'`)
        .join(',')})
```
It would be nice to have some kind of `mapArrayToSql(arr)` util to handle the `()` and stringification/join logic.

File: packages/mds-db/geographies.ts [hunk 13]
```diff
@@ -128,13 +145,16 @@ export async function editGeography(geography: Geography) {
     }
   })
   const sql = `UPDATE ${schema.TABLE.geographies} SET ${conditions} WHERE geography_id='${geography.geography_id}' AND publish_date IS NULL`
+
   await client.query(sql, vals.values())
-  return geography
+  const res = await readSingleGeography(geography.geography_id)
```
Comment (macsj200) [inline packages/mds-db/geographies.ts]: Nit: why not
```
return await readSingleGeography(geography.geography_id)
```

Comment (drtyh2o) [reply packages/mds-db/geographies.ts]: `return await` is redundant and will probably be flagged by the linter. You can just `return readSingleGeography(...)` though if you prefer.

File: packages/mds-db/geographies.ts [hunk 14]
```diff
@@ -205,3 +236,16 @@ export async function updateGeographyMetadata(geography_metadata: GeographyMetad
     ...recorded_metadata
   }
 }
+
+export async function deleteGeographyMetadata(geography_id: UUID) {
```
Comment (macsj200) [inline packages/mds-db/geographies.ts]: Should this `throw` if it doesn't find a corresponding `geography_id`?

File: packages/mds-db/tests/mds-db.spec.ts [hunk 15]
```diff
@@ -461,6 +476,15 @@ if (pg_info.database) {
         const policies = await MDSDBPostgres.findPoliciesByGeographyID(LAGeography.geography_id)
         assert.deepEqual(policies[0].policy_id, POLICY3_JSON.policy_id)
       })
+
+      it('throws if both get_published and get_unpublished are true for bulk geo reads', async () => {
+        await assert.rejects(
```
Comment (macsj200) [inline packages/mds-db/tests/mds-db.spec.ts]: I'm going to have to remember this pattern :D

File: packages/mds-db/tests/mds-db.spec.ts [hunk 16]
```diff
@@ -396,6 +399,18 @@ if (pg_info.database) {
         assert.deepEqual(unpublishedResult, false)
       })
 
+      it('.readGeographies understands all its parameters', async () => {
+        const publishedResult = await MDSDBPostgres.readGeographies({ get_published: true })
+        assert.deepEqual(publishedResult.length, 1)
+        assert.deepEqual(!!publishedResult[0].publish_date, true)
```
Comment (macsj200) [inline packages/mds-db/tests/mds-db.spec.ts]: Would this suffice?
```
assert.deepEqual(!publishedResult[0].publish_date, false)
```

File: packages/mds-geography-author/api.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (macsj200) [reply packages/mds-geography-author/api.ts]: We have very similar semantics in `packages/mds-metrics/api.ts`. We still follow the first point, but for the second instead of throwing we simply ignore the requested params in lieu of whatever is attached to the JWT `claim`.

...Perhaps we should change the metrics endpoint to throw as well if the client asks for something they're not allowed to see. @evanxadkins do you have any opinions on this?

File: packages/mds-db/geographies.ts [hunk 17]
```diff
@@ -10,35 +10,53 @@ import { getReadOnlyClient, getWriteableClient } from './client'
 import { ReadGeographiesParams, PublishGeographiesParams } from './types'
 
 export async function readSingleGeography(geography_id: UUID): Promise<Geography> {
-  try {
-    const client = await getReadOnlyClient()
+  const client = await getReadOnlyClient()
 
-    const sql = `select * from ${schema.TABLE.geographies} where geography_id = '${geography_id}'`
```
Comment (evanxadkins) [inline packages/mds-db/geographies.ts]: I believe we are trying to maintain the use of the sqlVals from the sql-utils library throughout our codebase.

File: packages/mds-geography-author/api.ts [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+
+  app.get(
+    pathsFor('/geographies/:geography_id'),
+    checkAccess(scopes => {
+      return scopes.includes('geographies:read:published') || scopes.includes('geographies:read:unpublished')
+    }),
+    async (req, res) => {
+      const { geography_id } = req.params
+      try {
+        const geography = await db.readSingleGeography(geography_id)
```
Comment (macsj200) [inline packages/mds-geography-author/api.ts]: What if we filtered the db query such that it won't return any `geography` without a `publish_date`?
That could make the app code simpler, but maybe it's important that we differentiate between "no geography found" vs. "scope doesn't permit reading this geography"

File: packages/mds-geography-author/api.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (drtyh2o) [reply packages/mds-geography-author/api.ts]: My general opinion is to filter out anything a user shouldn't see unless they specifically request something prohibited then in that case throw and return`403 - Forbidden`. Most common manifestation is when a user requests a list of things, filter and return what they can see and `200 - Ok` even if the list ends up being empty. If they request something by Id then throw/403 if access isn't allowed.

File: packages/mds-geography-author/api.ts [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+    }
+  )
+
+  app.delete(
+    pathsFor('/geographies/:geography_id'),
+    checkAccess(scopes => scopes.includes('geographies:write')),
+    async (req, res) => {
+      const { geography_id } = req.params
+      try {
+        await db.deleteGeographyMetadata(geography_id)
```
Comment (macsj200) [inline packages/mds-geography-author/api.ts]: If we change `deleteGeographyMetadata()` to `throw` if there's no associated `geography`, we could parallelize
```
await Promise.all([db.deleteGeographyMetadata(geography_id), db.deleteGeography(geography_id)])
```

File: packages/mds-geography-author/api.ts [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+  app.get(
+    pathsFor('/geographies'),
+    checkAccess(scopes => {
+      return scopes.includes('geographies:read:published') || scopes.includes('geographies:read:unpublished')
+    }),
+    async (req, res) => {
+      const summary = req.query.summary === 'true'
+      const { get_published = null, get_unpublished = null } = req.query
+      const params = { get_published, get_unpublished }
+      if (get_published) {
```
Comment (macsj200) [inline packages/mds-geography-author/api.ts]: I believe that `readGeographies()` does some validation such that both `get_published` and `get_unpublished` cannot both be set to `true`. Would it make sense to do the validation here, since this is where we parse the query params?

File: packages/mds-db/tests/mds-db.spec.ts (hunk 16 above)
Comment (janedotx) [reply packages/mds-db/tests/mds-db.spec.ts]: I like the comparison to `true` because it seems to me to better indicate that the `publish_date` is there.

File: packages/mds-geography-author/api.ts (hunk 20 above)
Comment (janedotx) [reply packages/mds-geography-author/api.ts]: You don't think that's kinda redundant? I'd rather throw just once in one layer.

File: packages/mds-db/tests/mds-db.spec.ts (hunk 16 above)
Comment (macsj200) [reply packages/mds-db/tests/mds-db.spec.ts]: That seems reasonable

File: packages/mds-geography-author/api.ts (hunk 20 above)
Comment (macsj200) [reply packages/mds-geography-author/api.ts]: I agree that all the params validation should happen in one place ideally, I just thought maybe it would make sense to validate the query params coming in from the request before we get into the `db` method, but I'm fine with either approach.

File: packages/mds-geography-author/api.ts (hunk 19 above)
Comment (janedotx) [reply packages/mds-geography-author/api.ts]: Isn't there the risk of a race where the geography and metadata are both present, but the geography gets deleted first, and then the metadata doesn't get deleted?

File: packages/mds-db/geographies.ts (hunk 14 above)
Comment (janedotx) [reply packages/mds-db/geographies.ts]: I don't think so. If there's a metadata without a geo, we should definitely be able to delete it.

File: packages/mds-geography-author/api.ts (hunk 19 above)
Comment (macsj200) [reply packages/mds-geography-author/api.ts]: I didn't consider that case, that sounds like it would cause some issues.

File: packages/mds-geography-author/LICENSE (hunk 5 above)
Comment (karcass) [reply packages/mds-geography-author/LICENSE]: Honestly I'm not sure whether we need them in some/all/just-the-root.  I'll ask @sarob.

File: packages/mds-db/tests/mds-db.spec.ts (hunk 16 above)
Comment (karcass) [reply packages/mds-db/tests/mds-db.spec.ts]: yeah I like `!!` better than `Boolean()` but probably the latter is more "official"

File: packages/mds-db/geographies.ts [hunk 21]
```diff
@@ -143,8 +163,12 @@ export async function deleteGeography(geography_id: UUID) {
   return geography_id
 }
 
-export async function publishGeography(params: PublishGeographiesParams) {
-  const { geography_id, publish_date } = params
+export async function publishGeography(params: PublishGeographiesParams): Promise<Geography> {
+  /* publish_date is a param instead of a default, because when a Policy is published,
```
Comment (avatarneil) [inline packages/mds-db/geographies.ts]: I wouldn't say we're using a param _instead_ of a default, considering we assign it a default value if none is passed.

File: packages/mds-db/geographies.ts [hunk 22]
```diff
@@ -157,51 +181,75 @@ export async function publishGeography(params: PublishGeographiesParams) {
     const conditions = []
     conditions.push(`publish_date = ${vals.add(publish_date)}`)
     const sql = `UPDATE ${schema.TABLE.geographies} SET ${conditions} where geography_id=${vals.add(geography_id)}`
-
     await client.query(sql, vals.values())
-    return geography_id
+    const res = await readSingleGeography(geography_id)
+    return res
```
Comment (avatarneil) [inline packages/mds-db/geographies.ts]: I think you could do this with one query by doing an `UPDATE ... returning *` or something along those lines, instead of updating and then querying.

File: packages/mds-db/geographies.ts (hunk 14 above)
Comment (avatarneil) [reply packages/mds-db/geographies.ts]: We should never have metadata without a corresponding geo though

File: packages/mds-db/tests/mds-db.spec.ts (hunk 15 above)
Comment (avatarneil) [reply packages/mds-db/tests/mds-db.spec.ts]: You can actually specify the full error object itself rather than just the name with `assert.throws`, check out our tests in `mds-schema-validators`

File: packages/mds-geography-author/api.ts [hunk 23]
```diff
@@ -0,0 +1,320 @@
...
+          /* No reason to let this bubble up. It's legit for metadata to not exist, and it
+           * seems wrong to throw an error for deleting metadata when this endpoint is mainly
+           * about deleting geographies.
+           */
+          await log.info(`Unable to delete nonexistent metadata for ${geography_id}`)
+        }
+        await db.deleteGeography(geography_id)
+        return res
+          .status(200)
+          .send({ result: `Successfully deleted geography and geography metadata of id ${geography_id}` })
```
Comment (avatarneil) [inline packages/mds-geography-author/api.ts]: Seems odd for the response to say we deleted geography_metadata even if it doesn't exist

File: packages/mds-db/geographies.ts (hunk 14 above)
Comment (janedotx) [reply packages/mds-db/geographies.ts]: I agree but shit happens @avatarneil

Comment (janedotx) [reply packages/mds-db/geographies.ts]: I agree but shit happens @avatarneil

Comment (janedotx) [reply packages/mds-db/geographies.ts]: I agree but shit happens @avatarneil

Comment (janedotx) [reply packages/mds-db/geographies.ts]: I agree but shit happens @avatarneil

Comment (janedotx) [reply packages/mds-db/geographies.ts]: I agree but shit happens @avatarneil

File: packages/mds-geography-author/LICENSE (hunk 5 above)
Comment (janedotx) [reply packages/mds-geography-author/LICENSE]: After talking to @karcass , it sounds like we can punt on the license issue for now.
